pkg/virt-operator/install-strategy/create.go

```diff
@@ -0,0 +1,995 @@
+	}
+
+	if !exists {
+		expectations.Deployment.RaiseExpectations(kvkey, 1, 0)
+		_, err = apps.Deployments(kv.Namespace).Create(deployment)
+		if err != nil {
+			expectations.Deployment.LowerExpectations(kvkey, 1, 0)
+			return fmt.Errorf("unable to create deployment %+v: %v", deployment, err)
+		}
+	} else if !objectMatchesVersion(&cachedDeployment.ObjectMeta, imageTag, imageRegistry) {
+		// Patch if old version
+		var ops []string
+
+		// Add Labels and Annotations Patches
+		labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&deployment.ObjectMeta)
+		if err != nil {
+			return err
+		}
+		ops = append(ops, labelAnnotationPatch...)
+
+		// Add Spec Patch
+		newSpec, err := json.Marshal(deployment.Spec)
+		if err != nil {
+			return err
+		}
+		ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec", "value": %s }`, string(newSpec)))
+
+		_, err = apps.Deployments(kv.Namespace).Patch(deployment.Name, types.JSONPatchType, generatePatchBytes(ops))
+		if err != nil {
+			return fmt.Errorf("unable to patch deployment %+v: %v", deployment, err)
+		}
+		log.Log.V(2).Infof("deployment %v updated", deployment.GetName())
+
+	} else {
+		log.Log.V(4).Infof("deployment %v is up-to-date", deployment.GetName())
+	}
+
+	return nil
+}
+
+func shouldTakeUpdatePath(targetVersion, currentVersion string) bool {
+
+	// if no current version, then this can't be an update
+	if currentVersion == "" {
+		return false
+	}
+
+	// semver doesn't like the 'v' prefix
+	targetVersion = strings.TrimPrefix(targetVersion, "v")
+	currentVersion = strings.TrimPrefix(currentVersion, "v")
+
+	// our default position is that this is an update.
+	// So if the target and current version do not
+	// adhere to the semver spec, we assume by default the
+	// update path is the correct path.
+	shouldTakeUpdatePath := true
+	target, err := semver.Make(targetVersion)
+	if err == nil {
+		current, err := semver.Make(currentVersion)
+		if err == nil {
+			if target.Compare(current) <= 0 {
+				shouldTakeUpdatePath = false
+			}
+		}
+	}
+
+	return shouldTakeUpdatePath
+}
+
+func SyncAll(kv *v1.KubeVirt,
+	prevStrategy *InstallStrategy,
+	targetStrategy *InstallStrategy,
+	stores util.Stores,
+	clientset kubecli.KubevirtClient,
+	expectations *util.Expectations) (bool, error) {
+
+	kvkey, err := controller.KeyFunc(kv)
+	if err != nil {
+		return false, err
+	}
+
+	gracePeriod := int64(0)
+	deleteOptions := &metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriod,
+	}
+
+	ext := clientset.ExtensionsClient()
+	core := clientset.CoreV1()
+	rbac := clientset.RbacV1()
+	scc := clientset.SecClient()
+
+	imageTag := kv.Status.TargetKubeVirtVersion
+	imageRegistry := kv.Status.TargetKubeVirtRegistry
+
+	takeUpdatePath := shouldTakeUpdatePath(kv.Status.TargetKubeVirtVersion, kv.Status.ObservedKubeVirtVersion)
+
+	// -------- CREATE AND ROLE OUT UPDATED OBJECTS --------
+
+	// create/update CRDs
+	for _, crd := range targetStrategy.crds {
```

Comment:
I think the flow for CRD updates should be the following:

1. check version bumps on all CRDs, break if not supported
2. Register a blocking webhook for all new CRDs (there needs to be no real webhook, not reachable would block)
3. Add new CRDs, so that our controllers can find it already when updating
4. update all the rest of the cluster
5. after virt-api is updated (which brings validations, defaulting and webhooks), update remaining CRDs
6. remove blocking webhooks for new CRDs.

Reply:
i agree. realistically this logic isn't going to make it for this pr.